File: java/src/main/java/S035.java
import com.google.common.collect.Sets;
import com.google.common.math.IntMath;

import java.math.RoundingMode;
import java.util.HashSet;
import java.util.Set;

public class S035 {

    private static final int LIMIT = 1000 * 1000;

    public static int solve() {
        HashSet<Integer> primes = Sets.newHashSet(Maths.generatePrimesBelow(LIMIT));
        return (int) primes.stream().filter(prime -> isCircular(primes, prime)).mapToInt(Integer::intValue).count();
    }

    private static boolean isCircular(Set<Integer> primes, int prime) {
        int digits = IntMath.log10(prime, RoundingMode.DOWN) + 1;
        int greatestDigit = IntMath.pow(10, digits - 1);
        int shifted = prime;
        for (int i = 0; i < digits - 1; i++) {
            int digit = shifted % 10;
            shifted /= 10;
            shifted += digit * greatestDigit;
            if (!primes.contains(shifted)) {
                return false;
            }
        }
        return true;
    }
}

File: cpp/S035.cpp
#include <iostream>
#include "primes.cpp"
#include "maths.cpp"

using namespace std;

static const int LIMIT = 1000 * 1000;

inline bool is_circular_prime(vector<bool>& sieve, int prime) {
    int rotation = prime;
    int pow_10 = POW_10[log_10(prime)];
    do {
        int digit = rotation % 10;
        if (!digit) {
            return false;
        }
        rotation  = (rotation / 10) + pow_10 * digit;
        if (!sieve[rotation]) {
            return false;
        }
    } while (rotation != prime);
    return true;
}

int main() {
    vector<bool> sieve = compute_sieve(LIMIT);
    int n_circular = 0;
    for (int i = 2; i < LIMIT; i++) {
        if (sieve[i]) {
            n_circular += is_circular_prime(sieve, i);
        }
    }
    cout << n_circular << endl;
    return 0;
}